lib/addgrps.c

```diff
@@ -14,29 +14,30 @@
 #include "prototypes.h"
 #include "defines.h"
 
-#include <stdio.h>
-#include <grp.h>
 #include <errno.h>
+#include <grp.h>
+#include <stdio.h>
+#include <string.h>
 
 #include "alloc/malloc.h"
 #include "alloc/reallocf.h"
 #include "shadowlog.h"
 
 #ident "$Id$"
 
-#define SEP ",:"
 /*
  * Add groups with names from LIST (separated by commas or colons)
  * to the supplementary group set.  Silently ignore groups which are
- * already there.  Warning: uses strtok().
+ * already there.
  */
-int add_groups (const char *list)
+int
+add_groups(const char *list)
 {
 	GETGROUPS_T *grouplist;
 	size_t i;
 	int ngroups;
 	bool added;
-	char *token;
+	char *g, *p;
```

Comment:
What does g and p mean? Can't we use token and subtoken or some other names?

Reply:
Since the function name is `add_groups`, `g` is of course a group.  :)

Also, it's used in a very short context:

```c
$ grepc add_groups lib/addgrps.c | grep '\<g\>' -C1
	bool added;
	char *g, *p;
	char buf[1024];
--
	p = buf;
	while (NULL != (g = strsep(&p, ",:"))) {
		struct group *grp;

		grp = getgrnam(g); /* local, no need for xgetgrnam */
		if (NULL == grp) {
			fprintf(shadow_logfd, _("Warning: unknown group %s\n"), g);
			continue;
```

The getgrnam(3) function accepts a group name as a string (`const char *`), so that confirms the intuitive idea that g is for group in a function that adds groups.

Reply:
`p` is just a short lived pointer:

```c
$ grepc add_groups lib/addgrps.c | grep '\<p\>' -C1
	bool added;
	char *g, *p;
	char buf[1024];
--
	added = false;
	p = buf;
	while (NULL != (g = strsep(&p, ",:"))) {
		struct group *grp;
```

Reply:
token is incorrect.  strsep(3) doesn't give tokens, but rather delimiter-separated values.  I consider token to be what strtok(3) gives, since it collapses contiguous delimiters (such as white-space in a C parser), and tokens must be non-empty.

Reply:
I have plans to make the function shorter after this patch, so that will hopefully add clarity.

Reply:
Then, can't we call them group and pointer? It's already a little bit complicated to understand the code as it stands now, I don't want to imagine in 5 or 10 years when several people will have touched it.

Reply:
I strongly oppose calling a variable `pointer` instead of `p`, in any context.  :-)

About `g` -> `group`, I could be convinced of that change, but I think it doesn't help at all in this context.  Consider that I'll apply #1123 to this function after this PR, and the function will end up being:

```c
$ grepc -tfd add_groups .
./lib/addgrps.c:int
add_groups(const char *list)
{
	char     *g, *p, *dup;
	FILE *shadow_logfd = log_get_logfd();
	gid_t    *gids;
	size_t   n;
	ssize_t  n0;

	dup = strdup(list);
	if (dup == NULL)
		return -1;

	n0 = getgroups(0, NULL);
	if (n0 == -1)
		goto free_dup;

	gids = MALLOC(n0, gid_t);
	if (gids == NULL)
		goto free_dup;

	n0 = getgroups(n0, gids);
	if (n0 == -1)
		goto free_gids;

	gids = REALLOCF(gids, n0 + strchrscnt(list, ",:") + 1, gid_t);
	if (gids == NULL)
		goto free_dup;

	n = n0;
	p = dup;
	while (NULL != (g = strsep(&p, ",:"))) {
		struct group  *grp;

		grp = getgrnam(g); /* local, no need for xgetgrnam */
		if (NULL == grp) {
			fprintf(shadow_logfd, _("Warning: unknown group %s\n"), g);
			continue;
		}

		LSEARCH(&grp->gr_gid, gids, &n);
	}

	if (setgroups(n, gids) == -1) {
		fprintf(shadow_logfd, "setgroups: %s\n", strerror(errno));
		goto free_gids;
	}

	free(gids);
	free(dup);
	return 0;

free_gids:
	free(gids);
free_dup:
	free(dup);
	return -1;
}
```

It becomes quite readable after all of those clean ups.  I don't think using `group` would help at all, and as Rob Pike's notes on programming in C say, it maybe

> obscures the details of the computation.  When the variable names are huge, it's harder to see what's going on.  This is partly a typographic issue;

@hallyn Would you like to untie?

Reply:
> @hallyn Would you like to untie?

This is of course subjective, but in general I would say a single letter variable is preferable if and only if its scope is small, say, < 10 lines.

This case is on the edge, let's leave it as is.  I think the proper fix would be to take the middle chunk of the function, the part using p and dup, and make it its own function :)  But that also is subjective.

Reply:
> I think the proper fix would be to take the middle chunk of the function, the part using p and dup, and make it its own function :) But that also is subjective.

You actually gave me an idea here.  And I'm doing quite a lot of progress there.

Here's a preview of what will come soon:

```c
alx@devuan:~/src/shadow/shadow/master$ grepc -tfd -h -B2 strsep2arr .
// strsep(3) a string into an array of strings.
// Return the number of fields in the string, or -1 on error.
inline ssize_t strsep2arr(char *s, const char *restrict delim,
    size_t n, char *a[restrict n])
{
	size_t  i;

	for (i = 0; i < n && s != NULL; i++)
		l[i] = strsep(&s, delim);

	if (s != NULL) {
		errno = E2BIG;
		return -1;
	}

	return i;
}
```
```c
alx@devuan:~/src/shadow/shadow/master$ grepc -tm -h STRSEP2ARR .
#define STRSEP2ARR(s, delim, a)                                       \
(                                                                     \
	strsep2arr(s, delim, NITEMS(a), a) == NITEMS(a) ? 0 : -1      \
)
```
```c
alx@devuan:~/src/shadow/shadow/master$ grepc -tfd -h -B1 strsep2ls .
// Like strsep2arr(), but add a NULL terminator.
inline ssize_t strsep2ls(char *s, const char *restrict delim,
    size_t n, char *ls[restrict n])
{
	size_t  i;

	i = strsep2arr(s, delim, n, ls);
	if (i >= n) {
		errno = E2BIG;
		return -1;
	}

	ls[i] = NULL;

	return i;
}
```
```c
alx@devuan:~/src/shadow/shadow/master$ grepc -tm -h STRSEP2LS .
#define STRSEP2LS(s, delim, ls)  strsep2ls(s, delim, NITEMS(ls), ls)
```

Reply:
I've just started using it in a few places, and I've already found a bug, and collapsed many lines of code:

```sh
$ git log -3 --oneline --stat
4745ef39 (HEAD -> astrsep2ls) lib/: Use STRSEP2ARR() instead of its pattern
 lib/gshadow.c       | 16 ++--------------
 lib/sgetpwent.c     | 20 +++++---------------
 lib/subordinateio.c | 18 ++++--------------
 3 files changed, 11 insertions(+), 43 deletions(-)
4f033a2e lib/port.c: getportent(): Use STRSEP2ARR() instead of its pattern
 lib/port.c | 22 +++++++---------------
 1 file changed, 7 insertions(+), 15 deletions(-)
38d0b90c lib/: Use STRSEP2LS() instead of its pattern
 lib/port.c | 17 +++--------------
 1 file changed, 3 insertions(+), 14 deletions(-)
```